File: cpp/_315.cpp
class FenwickTree {    
public:
    FenwickTree(int n): sums_(n + 1, 0) {}
    
    void update(int i, int delta) {
        while (i < sums_.size()) {
            sums_[i] += delta;
            i += lowbit(i);
        }
    }
    
    int query(int i) const {        
        int sum = 0;
        while (i > 0) {
            sum += sums_[i];
            i -= lowbit(i);
        }
        return sum;
    }
private:
    static inline int lowbit(int x) { return x & (-x); }
    vector<int> sums_;
};

// O(NlogN) / O(N)
class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        // compress
        set<int> st(nums.begin(), nums.end());
        unordered_map<int, int> rank;
        int i = 0;
        for (int num : st)
            rank[num] = ++i;
        
        int n = nums.size();
        vector<int> ans(n, 0);
        FenwickTree fw_tree(n);
        for (int i = n - 1; i >= 0; --i) {
            ans[i] = fw_tree.query(rank[nums[i]] - 1);
            fw_tree.update(rank[nums[i]], 1);
        }
        return ans;
    }
};

File: python/_315.py
class FenwickTree:
    def __init__(self, n):
        self.sums = [0] * (n + 1)
        
    def update(self, i, delta):
        while i < len(self.sums):
            self.sums[i] += delta
            i += self._lowbit(i)
    
    def query(self, i):
        s = 0
        while i > 0:
            s += self.sums[i]
            i -= self._lowbit(i)
        return s
        
    def _lowbit(self, x):
        return x & -x

# O(NlogN) / O(N)
class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        # compress
        all_nums = list(set(nums))
        rank = {val: i + 1 for i, val in enumerate(sorted(all_nums))}
        
        n = len(nums)
        ans = [0] * n
        tree = FenwickTree(n)
        for i in range(n - 1, -1, -1):
            ans[i] = tree.query(rank[nums[i]] - 1)
            tree.update(rank[nums[i]], 1)
        return ans